include/alpaka/meta/SyclPrintf.hpp

```diff
@@ -0,0 +1,35 @@
+/* Copyright 2022 Andrea Bocci
+ *
+ * This file is part of alpaka.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#pragma once
+
+#include <CL/sycl.hpp>
+
+#ifdef ALPAKA_ACC_SYCL_ENABLED
+
+// Kill printf in AMD GPU code because of missing compiler support
+#    ifdef __AMDGCN__
+#        include <cstdio> // the define breaks <cstdio> if it is included afterwards
+#        define printf(...)
+#    else
+
+#        ifdef __SYCL_DEVICE_ONLY__
+#            define CONSTANT __attribute__((opencl_constant))
+#        else
+#            define CONSTANT
+#        endif
+
+#        define printf(FORMAT, ...)                                                                                   \
+            do                                                                                                        \
+            {                                                                                                         \
+                static const CONSTANT char format[] = FORMAT;                                                         \
+                sycl::ext::oneapi::experimental::printf(format, ##__VA_ARGS__);                                       \
```

Comment:
According to [Intel's documentation](https://www.intel.com/content/www/us/en/developer/articles/technical/how-to-debug-gpu-device-code.html#1b) (it's not formatted well - scroll up to the section directly above) we need to make sure that the types are properly casted.

Reply:
I see what you mean.

Unfortunately I don't think that can be implemented in a macro ?

Anyway, it's not significantly different from the standard `printf`; that will also print garbage if the type specifiers are wrong:
```c++
#include <cstdio>

int main() {
  std::printf("this is not a float: %d\n", 3.14f);
}
```
compiles (with a warning) and runs, and prints (for example - it seems different every time):
```
this is not a float: -1967609000
```